pkg/analyzer/detection/detection.go

```diff
@@ -25,40 +26,63 @@ func DetectComponents(name string, files analyzer.Files, parent *database.Layer,
 	// If no CPEs are found on RHCOS, a note needs to be added that informs users of it.
 	// Also, the bool logic can be refactored into a single UseCertifiedWorkflow function.
 	// TODO(ROX-13906, ROX-14028): Implement note and refactor
-	if namespace != nil && (wellknownnamespaces.IsRHELNamespace(namespace.Name) || wellknownnamespaces.IsRHCOSNamespace(namespace.Name)) {
-		// This is a certified image that needs to be scanned differently.
-		// Use the RHELv2 scanner instead.
-		packages, cpes, err := rpm.ListFeatures(files)
+	if isCertifiedRHELNamespace(namespace) {
+		rhelFeatures, languageComponents, err = detectAndAnnotateCertifiedRHELComponents(name, files, namespace, languageComponents)
 		if err != nil {
 			return nil, nil, nil, nil, err
 		}
-		rhelfeatures = &database.RHELv2Components{
-			Dist:     namespace.Name,
-			Packages: packages,
-			CPEs:     cpes,
-		}
-		logrus.WithFields(logrus.Fields{LogLayerName: name, "rhel package count": len(packages), "rhel cpe count": len(cpes)}).Debug("detected rhelv2 features")
-		if err := rpm.AnnotateComponentsWithPackageManagerInfo(files, languageComponents); err != nil {
-			logrus.WithError(err).Errorf("Failed to analyze package manager info for language components: %s", name)
-		}
+		logrus.WithFields(logrus.Fields{
+			LogLayerName:         name,
+			"rhel package count": len(rhelFeatures.Packages),
+			"rhel cpe count":     len(rhelFeatures.CPEs),
+			"rhel content sets":  len(rhelFeatures.ContentSets),
+		}).Debug("detected rhelv2 features")
 	} else {
-		var err error
-		// Detect features.
 		featureVersions, err = detectFeatureVersions(name, files, namespace, parent)
 		if err != nil {
 			return nil, nil, nil, nil, err
 		}
-		if len(featureVersions) > 0 {
-			logrus.WithFields(logrus.Fields{LogLayerName: name, "feature count": len(featureVersions)}).Debug("detected features")
-		}
+		logrus.WithFields(logrus.Fields{
+			LogLayerName:    name,
+			"feature count": len(featureVersions),
+		}).Debug("detected features")
+	}
+	return namespace, featureVersions, rhelFeatures, languageComponents, nil
+}
+
+func isCertifiedRHELNamespace(namespace *database.Namespace) bool {
+	if namespace == nil {
+		return false
+	}
+	return wellknownnamespaces.IsRHELNamespace(namespace.Name) ||
+		wellknownnamespaces.IsRHCOSNamespace(namespace.Name)
+}
+
+func detectAndAnnotateCertifiedRHELComponents(name string, files analyzer.Files, namespace *database.Namespace, languageComponents []*component.Component) (*database.RHELv2Components, []*component.Component, error) {
+	// This is a certified image that needs to be scanned differently.
+	// Use the RHELv2 scanner instead.
+	packages, contentSets, err := rpm.ListFeatures(files)
+	if err != nil {
+		return nil, nil, err
+	}
+	rhelfeatures := &database.RHELv2Components{
+		Dist:     namespace.Name,
+		Packages: packages,
+		// CPEs are mapped and returned with content sets for backward compatibility.
```

Comment:
Needed for scanning images in OCR. Perhaps we can update that path, too, which would also eliminate the need to pass those over from Central -> Sensor -> Scanner in another PR, if it works out

Reply:
I see where you coming from. When looking at this patch, it is natural to try to eliminate CPE mapping in detection altogether. But the handling outside RHCOS scanning might be more complex.

I intentionally haven't investigated further. Perhaps database models need to be adjusted in Central Scanner, for example. In any case, I already created an issue ROX-14414 to track this, as we should consider it in a different context.